File: Practice/atcoder/sumitrust2019/src/___e.py
# なぜか途中からREなる...
# ので、Python3は捨ててPHPで解いた

n = int(input())
s = list(map(int, input().split()))
color = 0  # appeared_zero (0~2)
ans = 1
MOD = 1000000007
cur = [-1, -1, -1]
for c in s:
    if c == 0:
        cur[color] = 0
        color += 1
        ans %= MOD
    else:
        x = cur.count(c - 1)
        ans *= x
        ans %= MOD
        if c - 1 in cur:
            idx = cur.index(c - 1)
            cur[idx] += 1
        else:
            print("????????????")
            exit()
ans *= 3 if color == 0 else 3 if color == 1 else 6
ans %= MOD
print(ans)

File: Practice/atcoder/sumitrust2019/src/e.php
<?php
  $n = intVal(trim(fgets(STDIN)));
  $s = array_map("intVal", explode(" ", trim(fgets(STDIN))));
  $color = 0;
  $ans = 1;
  $MOD = 1000000007;
  $cur = [-1, -1, -1];

  if($s[0] != 0){
    echo 0 . PHP_EOL;
    exit();
  }
  foreach($s as $c){
    if($c == 0){
      if($color >= 3){
        echo 0 . PHP_EOL;
        exit;
      }
      $cur[$color] = 0;
      $color += 1;
      $ans %= $MOD;
    }else{
      $x = 0;
      foreach($cur as $curn){
        if($curn == $c-1){
          $x += 1;
        }
      }
      if($x == 0){
        echo 0 . PHP_EOL;
        exit();
      }
      $ans *= $x;
      $ans %= $MOD;
      $idx = array_search($c-1, $cur);
      $cur[$idx] += 1;
    }
  }

  if($color == 0){
    $ans *= 1;
  }elseif($color == 1){
    $ans *= 3;
  }else{
    $ans *= 6;
  }
  $ans %= $MOD;
  echo $ans . PHP_EOL;
?>